File: optimization/bench/c/bench.c
/**
 * HRP Benchmark — C implementation
 * Compile: gcc -O3 -o hrp_bench bench.c -lm
 */

#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <string.h>
#include <time.h>
#include <stdbool.h>

// ── Timing ──

static double now_us(void) {
    struct timespec ts;
    clock_gettime(CLOCK_MONOTONIC, &ts);
    return ts.tv_sec * 1e6 + ts.tv_nsec / 1e3;
}

// ── Generate synthetic prices ──

static void generate_prices(double *prices, int n, int days) {
    unsigned long long seed = 42;
    for (int i = 0; i < n; i++) {
        seed = seed * 6364136223846793005ULL + 1;
        double start = 100.0 + (seed % 900);
        prices[i * days] = start;
        double vol = 0.01 + (seed % 50) * 0.001;
        for (int t = 1; t < days; t++) {
            seed = seed * 6364136223846793005ULL + 1;
            double u = ((double)seed / (double)0xFFFFFFFFFFFFFFFFULL) - 0.5;
            double ret = vol * u * 0.816;
            prices[i * days + t] = prices[i * days + t - 1] * exp(ret);
        }
    }
}

// ── Log returns ──

static void log_returns(const double *prices, double *rets, int n, int days) {
    for (int i = 0; i < n; i++) {
        for (int t = 0; t < days - 1; t++) {
            rets[i * (days - 1) + t] = log(prices[i * days + t + 1] / prices[i * days + t]);
        }
    }
}

// ── Covariance matrix ──

static void cov_matrix(const double *rets, double *cov, int n, int T) {
    double *means = (double *)calloc(n, sizeof(double));
    for (int i = 0; i < n; i++) {
        for (int t = 0; t < T; t++) means[i] += rets[i * T + t];
        means[i] /= T;
    }
    for (int i = 0; i < n; i++) {
        for (int j = i; j < n; j++) {
            double s = 0;
            for (int t = 0; t < T; t++)
                s += (rets[i * T + t] - means[i]) * (rets[j * T + t] - means[j]);
            cov[i * n + j] = cov[j * n + i] = s / (T - 1);
        }
    }
    free(means);
}

// ── Correlation & Distance ──

static void corr_matrix(const double *cov, double *corr, int n) {
    double *stds = (double *)malloc(n * sizeof(double));
    for (int i = 0; i < n; i++) stds[i] = sqrt(cov[i * n + i]);
    for (int i = 0; i < n; i++)
        for (int j = 0; j < n; j++)
            corr[i * n + j] = (stds[i] > 0 && stds[j] > 0) ? cov[i * n + j] / (stds[i] * stds[j]) : 0;
    free(stds);
}

static void dist_matrix(const double *corr, double *dist, int n) {
    for (int i = 0; i < n * n; i++) {
        double v = (1.0 - corr[i]) / 2.0;
        dist[i] = sqrt(v > 0 ? v : 0);
    }
}

// ── Average linkage ──

typedef struct { int i, j; double dist; int size; } LinkageRow;

static void average_linkage(const double *dist_in, LinkageRow *Z, int n) {
    int cap = 2 * n;
    double *D = (double *)malloc(cap * cap * sizeof(double));
    bool *active = (bool *)calloc(cap, sizeof(bool));
    int *sizes = (int *)malloc(cap * sizeof(int));

    for (int i = 0; i < cap * cap; i++) D[i] = 1e18;
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) D[i * cap + j] = dist_in[i * n + j];
        active[i] = true;
        sizes[i] = 1;
    }

    for (int step = 0; step < n - 1; step++) {
        double minD = 1e18;
        int mi = 0, mj = 0;
        for (int i = 0; i < n + step; i++) {
            if (!active[i]) continue;
            for (int j = i + 1; j < n + step; j++) {
                if (!active[j]) continue;
                if (D[i * cap + j] < minD) { minD = D[i * cap + j]; mi = i; mj = j; }
            }
        }
        int nid = n + step;
        sizes[nid] = sizes[mi] + sizes[mj];
        Z[step] = (LinkageRow){ mi, mj, minD, sizes[nid] };

        for (int k = 0; k < nid; k++) {
            if (!active[k] || k == mi || k == mj) continue;
            double nd = (D[mi * cap + k] * sizes[mi] + D[mj * cap + k] * sizes[mj]) / sizes[nid];
            D[nid * cap + k] = nd;
            D[k * cap + nid] = nd;
        }
        D[nid * cap + nid] = 0;
        active[mi] = false;
        active[mj] = false;
        active[nid] = true;
    }
    free(D); free(active); free(sizes);
}

// ── Leaf order (iterative) ──

static void leaf_order(const LinkageRow *Z, int n, int *order, int *out_len) {
    int *stack = (int *)malloc(2 * n * sizeof(int));
    int sp = 0, cnt = 0;
    stack[sp++] = n + (n - 2); // root
    while (sp > 0) {
        int node = stack[--sp];
        if (node < n) { order[cnt++] = node; continue; }
        const LinkageRow *r = &Z[node - n];
        stack[sp++] = r->j;
        stack[sp++] = r->i;
    }
    *out_len = cnt;
    free(stack);
}

// ── Quasi-diag ──

static void quasi_diag(const double *cov, const int *order, double *qd, int n) {
    for (int i = 0; i < n; i++)
        for (int j = 0; j < n; j++)
            qd[i * n + j] = cov[order[i] * n + order[j]];
}

// ── HRP weights ──

static double cluster_var(const double *cov, const int *idx, int m, int n) {
    double v = 0;
    for (int a = 0; a < m; a++)
        for (int b = 0; b < m; b++)
            v += cov[idx[a] * n + idx[b]];
    return v / ((double)m * m);
}

static void hrp_weights(const double *covQ, int n, double *w) {
    for (int i = 0; i < n; i++) w[i] = 1.0;

    // BFS-style bisection
    typedef struct { int *idx; int len; } Segment;
    Segment *queue = (Segment *)malloc(2 * n * sizeof(Segment));
    int qh = 0, qt = 0;
    int *all = (int *)malloc(n * sizeof(int));
    for (int i = 0; i < n; i++) all[i] = i;
    queue[qt++] = (Segment){ all, n };

    while (qh < qt) {
        Segment seg = queue[qh++];
        if (seg.len <= 1) { free(seg.idx); continue; }
        int mid = seg.len / 2;

        double vL = cluster_var(covQ, seg.idx, mid, n);
        double vR = cluster_var(covQ, seg.idx + mid, seg.len - mid, n);
        double alpha = (1.0 / vL) / (1.0 / vL + 1.0 / vR);

        for (int i = 0; i < mid; i++) w[seg.idx[i]] *= alpha;
        for (int i = mid; i < seg.len; i++) w[seg.idx[i]] *= (1.0 - alpha);

        int *left = (int *)malloc(mid * sizeof(int));
        int *right = (int *)malloc((seg.len - mid) * sizeof(int));
        memcpy(left, seg.idx, mid * sizeof(int));
        memcpy(right, seg.idx + mid, (seg.len - mid) * sizeof(int));

        queue[qt++] = (Segment){ left, mid };
        queue[qt++] = (Segment){ right, seg.len - mid };
        free(seg.idx);
    }
    free(queue);

    double sum = 0;
    for (int i = 0; i < n; i++) sum += w[i];
    for (int i = 0; i < n; i++) w[i] /= sum;
}

// ── Format time ──

static void fmt_time(double us, char *buf) {
    if (us < 1000) sprintf(buf, "%6.0fµs", us);
    else if (us < 1e6) sprintf(buf, "%6.1fms", us / 1e3);
    else sprintf(buf, "%6.2fs ", us / 1e6);
}

// ── Main ──

static void bench(int n, int days) {
    double *prices = (double *)malloc(n * days * sizeof(double));
    generate_prices(prices, n, days);

    int T = days - 1;
    double *rets = (double *)malloc(n * T * sizeof(double));
    double t0 = now_us();
    log_returns(prices, rets, n, days);
    double t_ret = now_us() - t0;

    double *cov = (double *)malloc(n * n * sizeof(double));
    t0 = now_us();
    cov_matrix(rets, cov, n, T);
    double t_cov = now_us() - t0;

    double *corr = (double *)malloc(n * n * sizeof(double));
    corr_matrix(cov, corr, n);

    double *dist = (double *)malloc(n * n * sizeof(double));
    dist_matrix(corr, dist, n);

    LinkageRow *Z = (LinkageRow *)malloc((n - 1) * sizeof(LinkageRow));
    t0 = now_us();
    average_linkage(dist, Z, n);
    double t_link = now_us() - t0;

    int *order = (int *)malloc(n * sizeof(int));
    int olen;
    leaf_order(Z, n, order, &olen);

    double *qd = (double *)malloc(n * n * sizeof(double));
    t0 = now_us();
    quasi_diag(cov, order, qd, n);
    double t_qd = now_us() - t0;

    double *w = (double *)malloc(n * sizeof(double));
    t0 = now_us();
    hrp_weights(qd, n, w);
    double t_w = now_us() - t0;

    double total = t_ret + t_cov + t_link + t_qd + t_w;

    char s1[32], s2[32], s3[32], s4[32], s5[32], s6[32];
    fmt_time(t_ret, s1); fmt_time(t_cov, s2); fmt_time(t_link, s3);
    fmt_time(t_qd, s4); fmt_time(t_w, s5); fmt_time(total, s6);
    printf("  %6d │ %s │ %s │ %s │ %s │ %s │ %s\n", n, s1, s2, s3, s4, s5, s6);

    free(prices); free(rets); free(cov); free(corr);
    free(dist); free(Z); free(order); free(qd); free(w);
}

int main(void) {
    printf("╔═══════════════════════════════════════════════════════════════════╗\n");
    printf("║          HRP Benchmark — C (gcc -O3)                            ║\n");
    printf("╚═══════════════════════════════════════════════════════════════════╝\n");
    printf("  365 daily observations per asset\n\n");
    printf("  %6s │ %8s │ %8s │ %8s │ %8s │ %8s │ %8s\n",
        "N", "LogRet", "Cov", "Linkage", "QuasiD", "Weights", "TOTAL");
    printf("  ───────────────────────────────────────────────────────────────────\n");

    int sizes[] = {10, 25, 50, 100, 200, 500, 1000, 2000, 5000, 10000};
    for (int i = 0; i < 10; i++) bench(sizes[i], 365);
    return 0;
}

File: optimization/bench/zig/bench.zig
const std = @import("std");
const math = std.math;
const Allocator = std.mem.Allocator;

// ── Timing ──

fn now_us() f64 {
    const ts = std.time.nanoTimestamp();
    return @as(f64, @floatFromInt(ts)) / 1000.0;
}

// ── Generate synthetic prices ──

fn generatePrices(alloc: Allocator, n: usize, days: usize) ![][]f64 {
    var prices = try alloc.alloc([]f64, n);
    var seed: u64 = 42;
    for (0..n) |i| {
        prices[i] = try alloc.alloc(f64, days);
        seed = seed *% 6364136223846793005 +% 1;
        const start: f64 = 100.0 + @as(f64, @floatFromInt(seed % 900));
        prices[i][0] = start;
        const vol: f64 = 0.01 + @as(f64, @floatFromInt(seed % 50)) * 0.001;
        for (1..days) |t| {
            seed = seed *% 6364136223846793005 +% 1;
            const u: f64 = @as(f64, @floatFromInt(seed)) / @as(f64, @floatFromInt(std.math.maxInt(u64))) - 0.5;
            const ret: f64 = vol * u * 0.816;
            prices[i][t] = prices[i][t - 1] * @exp(ret);
        }
    }
    return prices;
}

// ── Log returns ──

fn logReturns(alloc: Allocator, prices: []const []const f64) ![][]f64 {
    const n = prices.len;
    var rets = try alloc.alloc([]f64, n);
    for (0..n) |i| {
        const days = prices[i].len;
        rets[i] = try alloc.alloc(f64, days - 1);
        for (0..days - 1) |t| {
            rets[i][t] = @log(prices[i][t + 1] / prices[i][t]);
        }
    }
    return rets;
}

// ── Covariance matrix (flat) ──

fn covMatrix(alloc: Allocator, rets: []const []const f64) ![]f64 {
    const n = rets.len;
    const T = rets[0].len;
    var means = try alloc.alloc(f64, n);
    defer alloc.free(means);
    for (0..n) |i| {
        var s: f64 = 0;
        for (0..T) |t| s += rets[i][t];
        means[i] = s / @as(f64, @floatFromInt(T));
    }
    var cov = try alloc.alloc(f64, n * n);
    for (0..n) |i| {
        for (i..n) |j| {
            var s: f64 = 0;
            for (0..T) |t| {
                s += (rets[i][t] - means[i]) * (rets[j][t] - means[j]);
            }
            const v = s / @as(f64, @floatFromInt(T - 1));
            cov[i * n + j] = v;
            cov[j * n + i] = v;
        }
    }
    return cov;
}

// ── Correlation & distance ──

fn corrMatrix(alloc: Allocator, cov: []const f64, n: usize) ![]f64 {
    var stds = try alloc.alloc(f64, n);
    defer alloc.free(stds);
    for (0..n) |i| stds[i] = @sqrt(cov[i * n + i]);
    var corr = try alloc.alloc(f64, n * n);
    for (0..n) |i| {
        for (0..n) |j| {
            corr[i * n + j] = if (stds[i] > 0 and stds[j] > 0)
                cov[i * n + j] / (stds[i] * stds[j])
            else
                0;
        }
    }
    return corr;
}

fn distMatrix(alloc: Allocator, corr: []const f64, n: usize) ![]f64 {
    var dist = try alloc.alloc(f64, n * n);
    for (0..n * n) |i| {
        const v = (1.0 - corr[i]) / 2.0;
        dist[i] = @sqrt(if (v > 0) v else 0);
    }
    return dist;
}

// ── Average linkage ──

const LinkageRow = struct { i: usize, j: usize, dist: f64, size: usize };

fn averageLinkage(alloc: Allocator, dist_in: []const f64, n: usize) ![]LinkageRow {
    const cap = 2 * n;
    var D = try alloc.alloc(f64, cap * cap);
    defer alloc.free(D);
    @memset(D, 1e18);
    var active = try alloc.alloc(bool, cap);
    defer alloc.free(active);
    @memset(active, false);
    var sizes = try alloc.alloc(usize, cap);
    defer alloc.free(sizes);

    for (0..n) |i| {
        for (0..n) |j| D[i * cap + j] = dist_in[i * n + j];
        active[i] = true;
        sizes[i] = 1;
    }

    var Z = try alloc.alloc(LinkageRow, n - 1);

    for (0..n - 1) |step| {
        var minD: f64 = 1e18;
        var mi: usize = 0;
        var mj: usize = 0;
        for (0..n + step) |i| {
            if (!active[i]) continue;
            for (i + 1..n + step) |j| {
                if (!active[j]) continue;
                if (D[i * cap + j] < minD) {
                    minD = D[i * cap + j];
                    mi = i;
                    mj = j;
                }
            }
        }
        const nid = n + step;
        sizes[nid] = sizes[mi] + sizes[mj];
        Z[step] = LinkageRow{ .i = mi, .j = mj, .dist = minD, .size = sizes[nid] };

        for (0..nid) |k| {
            if (!active[k] or k == mi or k == mj) continue;
            const nd = (D[mi * cap + k] * @as(f64, @floatFromInt(sizes[mi])) +
                D[mj * cap + k] * @as(f64, @floatFromInt(sizes[mj]))) /
                @as(f64, @floatFromInt(sizes[nid]));
            D[nid * cap + k] = nd;
            D[k * cap + nid] = nd;
        }
        D[nid * cap + nid] = 0;
        active[mi] = false;
        active[mj] = false;
        active[nid] = true;
    }
    return Z;
}

// ── Leaf order ──

fn leafOrder(alloc: Allocator, Z: []const LinkageRow, n: usize) ![]usize {
    var order = try alloc.alloc(usize, n);
    var stack = try alloc.alloc(usize, 2 * n);
    defer alloc.free(stack);
    var sp: usize = 0;
    var cnt: usize = 0;
    stack[sp] = n + Z.len - 1;
    sp += 1;
    while (sp > 0) {
        sp -= 1;
        const node = stack[sp];
        if (node < n) {
            order[cnt] = node;
            cnt += 1;
            continue;
        }
        const r = Z[node - n];
        stack[sp] = r.j;
        sp += 1;
        stack[sp] = r.i;
        sp += 1;
    }
    return order;
}

// ── Quasi-diag ──

fn quasiDiag(alloc: Allocator, cov: []const f64, order: []const usize, n: usize) ![]f64 {
    var qd = try alloc.alloc(f64, n * n);
    for (0..n) |i| {
        for (0..n) |j| {
            qd[i * n + j] = cov[order[i] * n + order[j]];
        }
    }
    return qd;
}

// ── HRP weights ──

fn hrpWeights(alloc: Allocator, covQ: []const f64, n: usize) ![]f64 {
    var w = try alloc.alloc(f64, n);
    @memset(w, 1.0);

    const Seg = struct { idx: []usize };
    var queue = try alloc.alloc(Seg, 2 * n);
    defer alloc.free(queue);
    var qh: usize = 0;
    var qt: usize = 0;

    var all = try alloc.alloc(usize, n);
    for (0..n) |i| all[i] = i;
    queue[qt] = Seg{ .idx = all };
    qt += 1;

    while (qh < qt) {
        const seg = queue[qh];
        qh += 1;
        if (seg.idx.len <= 1) { alloc.free(seg.idx); continue; }
        const mid = seg.idx.len / 2;

        var vL: f64 = 0;
        for (0..mid) |a| for (0..mid) |b| {
            vL += covQ[seg.idx[a] * n + seg.idx[b]];
        };
        vL /= @as(f64, @floatFromInt(mid * mid));

        var vR: f64 = 0;
        const rlen = seg.idx.len - mid;
        for (mid..seg.idx.len) |a| for (mid..seg.idx.len) |b| {
            vR += covQ[seg.idx[a] * n + seg.idx[b]];
        };
        vR /= @as(f64, @floatFromInt(rlen * rlen));

        const alpha = (1.0 / vL) / (1.0 / vL + 1.0 / vR);
        for (0..mid) |i| w[seg.idx[i]] *= alpha;
        for (mid..seg.idx.len) |i| w[seg.idx[i]] *= (1.0 - alpha);

        var left = try alloc.alloc(usize, mid);
        @memcpy(left, seg.idx[0..mid]);
        var right = try alloc.alloc(usize, rlen);
        @memcpy(right, seg.idx[mid..]);

        queue[qt] = Seg{ .idx = left };
        qt += 1;
        queue[qt] = Seg{ .idx = right };
        qt += 1;
        alloc.free(seg.idx);
    }

    var sum: f64 = 0;
    for (0..n) |i| sum += w[i];
    for (0..n) |i| w[i] /= sum;
    return w;
}

// ── Format ──

fn fmtTime(us: f64, buf: []u8) []u8 {
    if (us < 1000) {
        return std.fmt.bufPrint(buf, "{d:>6.0}µs", .{us}) catch buf[0..0];
    } else if (us < 1e6) {
        return std.fmt.bufPrint(buf, "{d:>6.1}ms", .{us / 1e3}) catch buf[0..0];
    } else {
        return std.fmt.bufPrint(buf, "{d:>6.2}s ", .{us / 1e6}) catch buf[0..0];
    }
}

fn benchOne(alloc: Allocator, n: usize, days: usize) !void {
    const prices = try generatePrices(alloc, n, days);
    defer {
        for (prices) |p| alloc.free(p);
        alloc.free(prices);
    }

    const prices_const: []const []const f64 = @ptrCast(prices);

    var t0 = now_us();
    const rets = try logReturns(alloc, prices_const);
    defer {
        for (rets) |r| alloc.free(r);
        alloc.free(rets);
    }
    const t_ret = now_us() - t0;

    const rets_const: []const []const f64 = @ptrCast(rets);

    t0 = now_us();
    const cov = try covMatrix(alloc, rets_const);
    defer alloc.free(cov);
    const t_cov = now_us() - t0;

    const corr = try corrMatrix(alloc, cov, n);
    defer alloc.free(corr);

    const dist = try distMatrix(alloc, corr, n);
    defer alloc.free(dist);

    t0 = now_us();
    const Z = try averageLinkage(alloc, dist, n);
    defer alloc.free(Z);
    const t_link = now_us() - t0;

    const order = try leafOrder(alloc, Z, n);
    defer alloc.free(order);

    t0 = now_us();
    const qd = try quasiDiag(alloc, cov, order, n);
    defer alloc.free(qd);
    const t_qd = now_us() - t0;

    t0 = now_us();
    const w = try hrpWeights(alloc, qd, n);
    defer alloc.free(w);
    const t_w = now_us() - t0;

    const total = t_ret + t_cov + t_link + t_qd + t_w;

    var b1: [32]u8 = undefined;
    var b2: [32]u8 = undefined;
    var b3: [32]u8 = undefined;
    var b4: [32]u8 = undefined;
    var b5: [32]u8 = undefined;
    var b6: [32]u8 = undefined;

    const stdout = std.io.getStdOut().writer();
    try stdout.print("  {d:>6} │ {s} │ {s} │ {s} │ {s} │ {s} │ {s}\n", .{
        n,
        fmtTime(t_ret, &b1),
        fmtTime(t_cov, &b2),
        fmtTime(t_link, &b3),
        fmtTime(t_qd, &b4),
        fmtTime(t_w, &b5),
        fmtTime(total, &b6),
    });
}

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    const alloc = gpa.allocator();

    const stdout = std.io.getStdOut().writer();
    try stdout.print("╔═══════════════════════════════════════════════════════════════════╗\n", .{});
    try stdout.print("║          HRP Benchmark — Zig (ReleaseFast)                      ║\n", .{});
    try stdout.print("╚═══════════════════════════════════════════════════════════════════╝\n", .{});
    try stdout.print("  365 daily observations per asset\n\n", .{});
    try stdout.print("  {s:>6} │ {s:>8} │ {s:>8} │ {s:>8} │ {s:>8} │ {s:>8} │ {s:>8}\n", .{
        "N", "LogRet", "Cov", "Linkage", "QuasiD", "Weights", "TOTAL",
    });
    try stdout.print("  ───────────────────────────────────────────────────────────────────\n", .{});

    const sizes = [_]usize{ 10, 25, 50, 100, 200, 500, 1000, 2000, 5000, 10000 };
    for (sizes) |n| {
        try benchOne(alloc, n, 365);
    }
}
